### PR TITLE
Disallow passing a empty array type to `abi` functions.

### LINKED ISSUE #13652: zero-length array types in expression context do not raise type errors
```
contract C {
  function f() public {
    abi.decode("",  (int[0][]));
  }
}
```

throws

https://github.com/ethereum/solidity/blob/84cdcec2cfb1fe9d4a9171d3ed0ffefd6107ee42/libsolidity/codegen/ABIFunctions.cpp#L1174

Repro

```
$ solc --ir test.sol
```

### PR DESCRIPTION
**Pull request by rodiazet**
This PR adds additional check in `TypeChecker` for `IndexAccess` which ensures that the array size is not `0`.
As a result of this change, type which is an empty array of contracts is also forbidden. 

Fixes https://github.com/argotorg/solidity/issues/13652.

Depends on: https://github.com/argotorg/solidity/pull/16490

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -22,10 +22,13 @@
  */
 
 #include <libsolidity/analysis/TypeChecker.h>
+
+#include "DeclarationTypeChecker.h"
```
Comment (stackenbotten3000) [inline libsolidity/analysis/TypeChecker.cpp]: Coding style error

**Comment by stackenbotten3000** [comment]
There was an error when running `chk_coding_style` for commit `aa31d42a9b1c270b3d5024b70e97d60c34f33da8`:
```
Coding style error:
libsolidity/analysis/TypeChecker.cpp:26:#include "DeclarationTypeChecker.h"

```
Please check that your changes are working as intended.

**Comment by stackenbotten3000** [comment]
There was an error when running `chk_coding_style` for commit `12c94081a054bcce42ac0aa36bf24a9360ee4bcc`:
```
Coding style error:
libsolidity/analysis/TypeChecker.cpp:26:#include "DeclarationTypeChecker.h"

```
Please check that your changes are working as intended.

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (stackenbotten3000) [inline libsolidity/analysis/TypeChecker.cpp]: Coding style error

**Review by nikola-matic — COMMENTED** [review]
Changelog

File: test/libsolidity/syntaxTests/array/invalid/library_index_access.sol [hunk 2]
```diff
@@ -1,6 +1,6 @@
 library C {
     function f() view public {
-        C[0];
+        C[1];
```
Comment (cameel) [inline test/libsolidity/syntaxTests/array/invalid/library_index_access.sol]: Can you also add new cases testing specifically for the case of zero length in a situation like this? Or do we have them already?

Comment (cameel) [reply test/libsolidity/syntaxTests/array/invalid/library_index_access.sol]: Also, I'd add variants for interfaces for completeness.

And rename `C` here to `L` (we usually use `C` for a contract so it makes it easy to confuse things).

File: test/libsolidity/syntaxTests/array/zero_length_array_type.sol
Comment (cameel) [inline test/libsolidity/syntaxTests/array/zero_length_array_type.sol:1]: This test is specifically about decoding so I'd rename it to `abi_decode_zero_length_array_type.sol`. You can have `int[0]` in other situations and this test is not checking them.

Comment (cameel) [reply test/libsolidity/syntaxTests/array/zero_length_array_type.sol:1]: BTW, I'm assuming we already have coverage for those other cases, but maybe check whether that's the case. These are ones I can think of:
- Variable declaration
    - Contract variable (`uint[0] x;`, `mapping(uint => uint[0]) m;`, `function() returns (uint[0] memory) f;`)
    - Local variable (`uint[0] storage x;`)
    - Function arguments and returns
    - Try/catch statement (`try this.g() returns (uint[0] memory) {} catch (bytes memory b) {}`)
- Tuple declaration (`(uint[0] memory a, uint b) = ([], 1);`)
- Struct field (`struct S { uint[0] x; }`)
- Error and event parameters
- Array allocation (`new uint[0][](3); }`)
- No-op statement consisting of just the type (`uint[0];`)

Technically, these are disallowed for various other reasons in analysis, but are still valid syntax, so would not hurt to have them covered:
- UDVT declaration (`type U is uint[0];`)
- `type()` builtin (`type(uint[0])`).
- Ternary operator (`true ? uint[0] : uint[0]`)
- Inline Array indexing (`[uint[0]][0]`)

File: test/libsolidity/syntaxTests/array/zero_length_array_type.sol [hunk 3]
```diff
@@ -0,0 +1,38 @@
+contract C {
+    function f0() public {
+        abi.decode("",  (int[0]));
```
Comment (cameel) [inline test/libsolidity/syntaxTests/array/zero_length_array_type.sol]:
```suggestion
        abi.decode("", (int[0]));
```

File: test/libsolidity/syntaxTests/array/zero_length_array_type.sol
Comment (cameel) [reply test/libsolidity/syntaxTests/array/zero_length_array_type.sol:1]: Oh, I see `f6()` is actually not decoding. I'd still split it into separate tests though.

Comment (cameel) [inline test/libsolidity/syntaxTests/array/zero_length_array_type.sol:1]: Please also add the case from https://github.com/argotorg/solidity/issues/13652#issuecomment-1292287194. That one was not a compilation error so it's important that this no longer allowed:

```solidity
contract C {
    function f() public returns (bytes memory) {
        return abi.encode(abi.decode("", (int[0])));
    }
}
```

Comment (cameel) [inline test/libsolidity/syntaxTests/array/zero_length_array_type.sol:1]: Please also add a dedicated case for `super` (which is actually a type):
```solidity
contract C {
    function f() public {
        super[0];
    }
}
```

```solidity
contract C {
    function f() public {
        super[0] s;
    }
}
```

Comment (cameel) [reply test/libsolidity/syntaxTests/array/zero_length_array_type.sol:1]: Interestingly, an array of `super` is a valid expression on its own (`super[3];`) even though you cannot use it in a variable declaration (`super[3] s;`). I think it should be disallowed in either case.

I suspect that we have more things like this. Would be good to take a look at `DeclarationTypeChecker` and compare what other checks it has for `ArrayTypeName` that we do not here. Perhaps we should extract them into a shared function to avoid such discrepancies in the future.

Comment (cameel) [reply test/libsolidity/syntaxTests/array/zero_length_array_type.sol:1]: Also, pinging @matheusaaguiar to pay attention to this in test cases related to comptime. Looks like we have two places in the code where we evaluate array sizes. These should always behave the same way.

Also, WTF, why does every random issue recently turn out to be related to comptime one way or another :D

Comment (rodiazet) [reply test/libsolidity/syntaxTests/array/zero_length_array_type.sol:1]: Do not open your refrigerator.

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 4]
```diff
@@ -18,7 +18,7 @@
 
 #include <libsolidity/analysis/DeclarationTypeChecker.h>
 
-#include <libsolidity/analysis/ConstantEvaluator.h>
+#include "TypeChecker.h"
```
Comment (stackenbotten3000) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]: Coding style error

File: libsolidity/analysis/TypeChecker.cpp [hunk 5]
```diff
@@ -22,10 +22,13 @@
  */
 
 #include <libsolidity/analysis/TypeChecker.h>
+
+#include "ConstantEvaluator.h"
```
Comment (stackenbotten3000) [inline libsolidity/analysis/TypeChecker.cpp]: Coding style error

File: libsolidity/analysis/TypeChecker.cpp [hunk 6]
```diff
@@ -3470,29 +3473,56 @@ bool TypeChecker::visit(IndexAccess const& _access)
 	case Type::Category::TypeType:
 	{
 		TypeType const& typeType = dynamic_cast<TypeType const&>(*baseType);
-		if (auto const* contractType = dynamic_cast<ContractType const*>(typeType.actualType()))
+		const auto actualType = typeType.actualType();
```
Comment (stackenbotten3000) [inline libsolidity/analysis/TypeChecker.cpp]: Coding style error

File: libsolidity/analysis/TypeChecker.cpp [hunk 7]
```diff
@@ -3470,29 +3473,56 @@ bool TypeChecker::visit(IndexAccess const& _access)
...
+		const auto actualTypeCategory = actualType->category();
```
Comment (stackenbotten3000) [inline libsolidity/analysis/TypeChecker.cpp]: Coding style error

**Comment by nikola-matic** [comment]
Run `./scripts/check_style.sh` before pushing to avoid style errors.

File: test/libsolidity/syntaxTests/array/invalid/library_index_access.sol (hunk 2 above)
Comment (rodiazet) [reply test/libsolidity/syntaxTests/array/invalid/library_index_access.sol]: Done

File: test/libsolidity/syntaxTests/array/zero_length_array_type.sol
Comment (rodiazet) [reply test/libsolidity/syntaxTests/array/zero_length_array_type.sol:1]: Reworked and more tests added.

File: test/libsolidity/syntaxTests/array/zero_length_array_type.sol (hunk 3 above)
Comment (rodiazet) [reply test/libsolidity/syntaxTests/array/zero_length_array_type.sol]: Fixed

File: test/libsolidity/syntaxTests/array/zero_length_array_type.sol
Comment (rodiazet) [reply test/libsolidity/syntaxTests/array/zero_length_array_type.sol:1]: Added as a separated test case.

Comment (rodiazet) [reply test/libsolidity/syntaxTests/array/zero_length_array_type.sol:1]: I reworked this part and put common logic in separated function, where it was possible.

File: libsolidity/analysis/TypeChecker.cpp [hunk 8]
```diff
@@ -3470,29 +3470,56 @@ bool TypeChecker::visit(IndexAccess const& _access)
 	case Type::Category::TypeType:
 	{
 		TypeType const& typeType = dynamic_cast<TypeType const&>(*baseType);
-		if (auto const* contractType = dynamic_cast<ContractType const*>(typeType.actualType()))
+		const auto actualType = typeType.actualType();
```
Comment (stackenbotten3000) [inline libsolidity/analysis/TypeChecker.cpp]: Coding style error

File: libsolidity/analysis/TypeChecker.cpp [hunk 9]
```diff
@@ -3470,29 +3470,56 @@ bool TypeChecker::visit(IndexAccess const& _access)
...
+		const auto actualTypeCategory = actualType->category();
```
Comment (stackenbotten3000) [inline libsolidity/analysis/TypeChecker.cpp]: Coding style error

Comment (rodiazet) [reply libsolidity/analysis/TypeChecker.cpp]: This error is not issued when running the script locally.

Comment (rodiazet) [reply libsolidity/analysis/TypeChecker.cpp]: AI says so:
```
With git grep -E, the regex engine is POSIX extended regex. In that dialect, \s is not a whitespace escape (that’s a PCRE/Perl thing). So:
^\s* is effectively interpreted as ^s* (i.e., “zero or more s characters at BOL”)
which means it won’t match indentation made of spaces/tabs
```

Comment (nikola-matic) [reply libsolidity/analysis/TypeChecker.cpp]: As usual, AI is full of shit. You're getting the error because you're using `const auto ...` instead of `auto const ...`. What OS are you running locally?

Comment (rodiazet) [reply libsolidity/analysis/TypeChecker.cpp]: I know what was the problem, but this script doesn't find the pattern. And this is good explanation. I'm using MacOS with git in the newest version `2.52.0`.

Comment (rodiazet) [reply libsolidity/analysis/TypeChecker.cpp]: AI proposed to change `\s`to `[[:space:]]`, because `(POSIX ERE) does not understand \s` :)

Looks like it's right. https://en.wikibooks.org/wiki/Regular_Expressions/POSIX-Extended_Regular_Expressions

Comment (rodiazet) [reply libsolidity/analysis/TypeChecker.cpp]: https://github.com/argotorg/solidity/pull/16441

Comment (rodiazet) [reply libsolidity/analysis/TypeChecker.cpp]: ~On the other hand it looks like they should be supported too https://www.boost.org/doc/libs/1_71_0/libs/regex/doc/html/boost_regex/syntax/basic_extended.html#boost_regex.syntax.basic_extended.single_character_character_class~
~Probable for some reason MacOS does not support it.~
It's for `boost.regexp` only

**Comment by cameel** [comment]
> [rodiazet](https://github.com/rodiazet) closed this in #16441 [39 minutes ago](https://github.com/argotorg/solidity/pull/16416#event-22592397388)

Hmm... looks like `Fixes <comment URL>` does not work the way you think it does :P

File: libsolidity/analysis/DeclarationTypeChecker.cpp [hunk 10]
```diff
@@ -335,36 +333,8 @@ void DeclarationTypeChecker::endVisit(ArrayTypeName const& _typeName)
...
-				length->location(),
-				"Array length too large, maximum is 2**256 - 1."
-			);
-
-		_typeName.annotation().type = TypeProvider::array(
-			DataLocation::Storage,
-			baseType,
-			lengthValue ? u256(lengthValue->numerator()) : u256(0)
-		);
+		auto const lengthValue = TypeChecker::checkArrayLengthExpression(*length, m_errorReporter);
```
Comment (cameel) [inline libsolidity/analysis/DeclarationTypeChecker.cpp]:
```suggestion
		u256 const lengthValue = TypeChecker::checkArrayLengthExpression(*length, m_errorReporter);
```

Comment (cameel) [reply libsolidity/analysis/DeclarationTypeChecker.cpp]: It would also be clearer to rename `length` to `lengthExpression` and for this one to be `length`.

File: libsolidity/analysis/TypeChecker.cpp [hunk 11]
```diff
@@ -4239,3 +4266,36 @@ bool TypeChecker::useABICoderV2() const
 	return *m_currentSourceUnit->annotation().useABICoderV2;
 
 }
+
+u256 TypeChecker::checkArrayLengthExpression(Expression const& _arrayLengthExpression, ErrorReporter& _errorReporter)
+{
+	std::optional<rational> lengthValue;
+	if (_arrayLengthExpression.annotation().type && _arrayLengthExpression.annotation().type->category() == Type::Category::RationalNumber)
+		lengthValue = dynamic_cast<RationalNumberType const&>(*_arrayLengthExpression.annotation().type).value();
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: Can you try removing this special case? I'm pretty sure `ConstantEvaluator` can already handle rational literals. I'm not sure why this is singled out here.

https://github.com/argotorg/solidity/blob/80d5c5362b5d41697fd62ef36cf6abd8d5a68cc4/libsolidity/analysis/ConstantEvaluator.cpp#L392-L396

File: libsolidity/analysis/TypeChecker.cpp [hunk 12]
```diff
@@ -4239,3 +4266,36 @@ bool TypeChecker::useABICoderV2() const
...
+		_errorReporter.typeError(
+			1847_error,
+			_arrayLengthExpression.location(),
+			"Array length too large, maximum is 2**256 - 1."
+		);
+	else
+		return u256(lengthValue->numerator());
+
+	solAssert(lengthValue || _errorReporter.hasErrors(), "Must have reported errors.");
+	return 0;
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: It the function does not always return a valid value, it should be returning an `optional` or throw. It's too easy for someone to forget to check the error reporter and use the `0` as the actual result.

File: libsolidity/analysis/TypeChecker.h [hunk 13]
```diff
@@ -59,6 +59,10 @@ class TypeChecker: private ASTConstVisitor
 
 	static bool typeSupportedByOldABIEncoder(Type const& _type, bool _isLibraryCall);
 
+	/// Checks that `_arrayLengthExpression` is allowed for array length value.
+	/// @returns a value of length or report an error.
+	static u256 checkArrayLengthExpression(Expression const& _arrayLengthExpression, langutil::ErrorReporter& _errorReporter);
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.h]: The fact that constant evaluation happens here is important enough that I'd include it in the name.

```suggestion
	/// Evaluates `_arrayLengthExpression` and checks if the value is allowed as an array length.
	/// @returns length if it can be evaluated without errors. Otherwise returns 0 and adds an error to the error reporter.
	static u256 evaluateAndCheckArrayLengthExpression(Expression const& _arrayLengthExpression, langutil::ErrorReporter& _errorReporter);
```

Alternatively, if my suggestion to remove the special case in evaluation works, you could keep the name if you do evaluation outside and have the function accept `TypedRational`.

Comment (cameel) [reply libsolidity/analysis/TypeChecker.h]: Also, I wonder if it would not make more sense to just put this function in `ConstantEvaluator`. It's used by two analysis components and the concept of array length is general enough that I think it would not be out of place. And `ConstantEvaluator` already uses an error reporter to report other errors.

File: libsolidity/analysis/TypeChecker.cpp [hunk 14]
```diff
@@ -3470,29 +3470,56 @@ bool TypeChecker::visit(IndexAccess const& _access)
 	case Type::Category::TypeType:
 	{
 		TypeType const& typeType = dynamic_cast<TypeType const&>(*baseType);
-		if (auto const* contractType = dynamic_cast<ContractType const*>(typeType.actualType()))
+		auto const actualType = typeType.actualType();
+		auto const actualTypeCategory = actualType->category();
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: You dropped the pointer. Also, this is more readable when the types are not hidden. They are very short.

```suggestion
		Type const* actualType = typeType.actualType();
		Category const actualTypeCategory = actualType->category();
```

Comment (cameel) [reply libsolidity/analysis/TypeChecker.cpp]: TBH I would not even introduce variables for these. `typeType->actualType()` and `typeType->actualType()->category()` are perfectly clear and not that long. On the other hand extra local variables always increase cognitive load because you have to keep track of that context in your head now.

File: libsolidity/analysis/TypeChecker.cpp [hunk 15]
```diff
@@ -3470,29 +3470,56 @@ bool TypeChecker::visit(IndexAccess const& _access)
...
 			{
-				if (auto indexValue = dynamic_cast<RationalNumberType const*>(type(*index)))
-					length = indexValue->literalValue(nullptr);
-				else
-					m_errorReporter.fatalTypeError(3940_error, index->location(), "Integer constant expected.");
+				m_errorReporter.typeError(
+					5530_error,
+					_access.location(),
+					"Index notation is not allowed for type.");
 			}
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: I suggested that we should disallow things like `super[3]`, but I would not do it right in this PR. This is unrelated to the fix and needs a separate changelog entry and tests. And we should first consider whether the change is breaking.

Can you file an issue first? I already suggested splitting off the refactor into a separate PR below and I don't want us to get multiple levels deep :) This bit can be left for later.

File: libsolidity/analysis/TypeChecker.cpp [hunk 16]
```diff
@@ -3470,29 +3470,56 @@ bool TypeChecker::visit(IndexAccess const& _access)
...
+		case Type::Category::FixedPoint:
+		case Type::Category::ArraySlice:
+		case Type::Category::Function:
+		case Type::Category::Tuple:
+		case Type::Category::TypeType:
+		case Type::Category::Modifier:
+		case Type::Category::Magic:
+		case Type::Category::Module:
+		case Type::Category::InaccessibleDynamic:
+			solAssert(false, "Unexpected actual type category of TypeType.");
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp]: Your message is more precise, but I think this one would give more context as to what went wrong:

```suggestion
			solAssert(false, "Unexpected type of array size expression.");
```

Still, it's just an assert so a message is optional.

File: libsolidity/analysis/TypeChecker.cpp [hunk 17]
```diff
@@ -3470,29 +3470,56 @@ bool TypeChecker::visit(IndexAccess const& _access)
...
-				length
-			));
+		if (!index)
+			resultType = TypeProvider::typeType(TypeProvider::array(DataLocation::Memory, actualType));
+		else
+		{
+			index->accept(*this);
+			auto const lengthValue = checkArrayLengthExpression(*index, m_errorReporter);
+			resultType = TypeProvider::typeType(TypeProvider::array(DataLocation::Memory, actualType, lengthValue));
 		}
```
Comment (cameel) [inline libsolidity/analysis/TypeChecker.cpp:3526]: Looks like we weren't using constant evaluator here previously, so this change actually affects functionality. Now we support a much wider range of expressions here. This needs to be included in a changelog entry and also needs test coverage.

I think that we might be better off splitting off this refactor into a separate PR and putting the fix on top of it, to make it clear which changes belong to which.

Comment (cameel) [reply libsolidity/analysis/TypeChecker.cpp:3526]: Alternatively, if you want to avoid a refactor PR, you could also just not use constant evaluator here so that the current behavior is preserved. We should be doing that eventually, but for the purpose of this fix I only wanted the checks to be unified. You can achieve that by keeping the evaluation out of `checkArrayLengthExpression()` like I suggested in https://github.com/argotorg/solidity/pull/16416#discussion_r2846381693.

In that case please file an issue for this though (i.e. for the fact that the types checked here do not support the same range of expressions as the ones in declarations; e.g. no constants or arithmetic on literals).

File: libsolidity/analysis/TypeChecker.cpp (hunk 15 above)
Comment (rodiazet) [reply libsolidity/analysis/TypeChecker.cpp]: https://github.com/argotorg/solidity/issues/16489

File: libsolidity/analysis/TypeChecker.cpp (hunk 17 above)
Comment (rodiazet) [reply libsolidity/analysis/TypeChecker.cpp:3526]: Do you mean to move the common code from `DeclarationTypeChecker` to `ConstantEvaluator` and don't change the `TypeChecker` logic, and later in the fix PR just use `evaluateAndCheckArrayLengthExpression` in type checker as well?
EDIT: I haven't seen @cameel last message when writing this one.

Comment (rodiazet) [reply libsolidity/analysis/TypeChecker.cpp:3526]: I will make the rector PR. I want to assure that the index expression in this two cases have the same semantic.

Comment (rodiazet) [reply libsolidity/analysis/TypeChecker.cpp:3526]: https://github.com/argotorg/solidity/pull/16490

Comment (cameel) [reply libsolidity/analysis/TypeChecker.cpp:3526]: > Do you mean to move the common code from `DeclarationTypeChecker` to `ConstantEvaluator` and don't change the `TypeChecker` logic, and later in the fix PR just use `evaluateAndCheckArrayLengthExpression` in type checker as well?

I meant move the common code to `evaluateAndCheckArrayLengthExpression()`, but without the zero-length check. Leave that last bit for the fix PR.

Though looking at it again now, maybe it would make more sense to do it the other way around. I.e. for the fix just duplicate the zero length check, without moving any code to a shared function. Then put the refactor on top of the fix and do the function there. The upside of that is that the fix is almost done so we'll probably be able to merge it much quicker than the refactor.

And don't forget that the "refactor" is not really just a refactor after all because it does affect the language. So it needs tests for the behavior that gets changed and a changelog entry.

Comment (rodiazet) [reply libsolidity/analysis/TypeChecker.cpp:3526]: PR is ready. Zero length check is only where it was initially, so in `DeclarationTypeChecker`. I haven't changed the logic of the `TypeChecker`. It's also only refactoring, so I haven't used `evaluateAndCheckArrayLengthExpression` there yet. On the top of it would make the fix PR which will only be using common function in the `TypeChecker` and all the additional required steps you mentioned. Tests + changelog.
